docs/IR.md

```diff
@@ -247,10 +249,16 @@ The representation distinguishes between two kinds of values: attribute values,
 
 The types of the inputs and outputs of a graph must be specified.
 
-#### Optional Inputs
+#### Variadic Inputs and Outputs
+ 
+The last input or output of an operator MAY be marked as variadic. For each variadic operator input, one or more node inputs must be specified. For each variadic operator output, one or more node outputs must be specified. For example, the operator 'Max()' can be used to compute the maximum of a varying number of input values.
+
+#### Optional Inputs and Outputs
 
 Some operators have inputs that are marked as optional. There are two ways to leave an optional input unspecified. The first is to simply not provide that input. However, this is not always possible - for example, if you wish to leave the fourth input unspecified, but still provide a value for the fifth input. Therefore, any input with a name of the empty string is treated as an unspecified optional input.
 
+Some operators have outputs that are optional. This means that an operator, depending on its input parameters and/or attributes, may not compute the corresponding value. Each node referring to such an operator must have knowledge of the circumstances under which output is computed or not, and provide a name for the output when it is computed. Failure to do so will result in the output not being captured.
```

Comment:
Consider LSTM with output_sequence=False as an example. Does "provide a name for the output when it is computed" mean that we can assign an empty string to LSTM's first output in this case?

Reply:
Optional outputs is perhaps the murkiest part of the semantics, hardest to explain, I think. 

Optional inputs make sense -- they don't have to be provided? 

Optional outputs, however -- who is it that has an option, the caller or the operator? In other words, is it that the operator has the option of not computing the value, or the model graph that has the option of not picking it up? I believe the former is what's I've heard described, the operator may not compute it. 

If that is correct, then looking at my wording above, the text is focusing on the wrong issue -- it's not about missing to pick up a value that is computed, the issue is giving a name to an output that is not actually computed, and then using it as input to something else.

Reply:
For LSTM/RNN/GRU, I think the existence of optional outputs (or inputs) means that an operator can have multiple static configurations because whether the first output is computed or not is determined by a user-specified attribute. Therefore, the user of such an operator should be in charge of choosing the right outputs. If she turns output_sequence on, the first output name would be a meaningful string. If not, she can use an empty string. My point is that as long as an output configuration is (implicitly) determined by an user, using an empty string is better than setting up a flag+provide an uncomputed tensor name.

Reply:
Yes, the optional outputs part is somewhat murky. My interpretation is that

(a) Generic (operator-independent) constraints: The model may omit an output (using an empty string or completely omitting it in the case of trailing outputs) only if that output is marked as optional in the operator specification.

(b) There will typically be further operator-specific constraints on when omitting an output is valid, which is currently captured only in the informal operator description.

Reply:
Regarding @NiklasGustafsson 's question as to who has the option: I believe it is a contract between both the caller and the operator, which unfortunately is only informally captured in the operator description. Since this part is not formally captured, "model validation" cannot check these aspects of the contract (e.g., such as "an output is computed iff a specific attribute has a specific value"). This, in turn, means that the semantics of what happens during a model's execution if this contract is violated is undefined.

Reply:
This raises a question -- if a node fails to give a name to an output that is computed (optional or not), what is the failure mode? It seems to me that giving a name to an output that is __not__ computed is definitely an error, while the reverse isn't as obviously wrong. I guess that some implementations could have trouble with both.

Reply:
@NiklasGustafsson , what do you mean by failure mode? Is it "missing output tensor name"? Also, why assigning an output name to an uncomputed output is wrong? I think it's fine if no other operator refers to that undefined output. Do you mean "forgetting assigning a name to a computed output is wrong"?

Reply:
Giving a name to an output that is not computed: this could happen even with non-optional outputs. When this named output is subsequently used in a different node, it would pretty much behave like the use of an uninitialized variable (the corresponding tensor memory could have random values). (Obviously, this depends on the runtime implementation.) Unfortunately, we cannot always catch this statically.

Failing to give a name for an output that is not optional is easier to handle: model validation can catch this, and a runtime can avoid initiating execution. Even when it can't be caught in model validation, it can be caught at runtime. How this exception is reported to the user is implementation-dependent. I guess ONNX could try to standardize the exception behavior in such cases.

Reply:
By "failure mode," I mean how does an error manifest itself at runtime? 

If an output (optional or not) is not named, it cannot be used as an input or graph output. What downstream error can arise from that? (I'm not saying there can't be one, it's just not clear to me what it could be.)

By assigning a name to an output that is __not__ computed, errors can arise, depending on the runtime implementation. Since such a name can be used as input to other operators, or as a graph output, uninitialized values may be propagated, which leads to further errors and, in the worst case, crashes. Failing to name an output that is computed could be an indication that the model is incorrectly constructed, which is worth at least a warning during validation, but it doesn't seem to be able to cause corruption of data, and if a model doesn't need the output, then why make it assign a name?

Just asking questions...

Reply:
Yes, for the second part, we can naturally extend the language to allow callers to indicate that they don't care about a particular output, e.g., perhaps by using a special name like "_". (But my understanding is that this scenario is not what the current optional-output mechanism targets.)

Reply:
OK, I'll change the text to reflect this.

Reply:
I think there are two different questions. Below I summarize our thread.

1. How to specify an operator's optional output when it's optional? I assume we all agree that it should be an empty string. If my assumption is correct, why do we have output_sequence flag in LSTM/GRU/RNN? It looks a bit redundant to me because it's implicitly determined by the first output's name (empty string --> output_sequence=False, meaningful string --> output_sequence=True). 

2. Is it legal if one doesn't provide a name (to the operator) for a computed output? Yes, if it's optional (equivalent to assigning an empty string). No, if it's not optional. @NiklasGustafsson

Reply:
Let me add a table for all possible output configurations. We can use it for further discussion.

| Optional  | Named (non-empty name) |  Computed  |  Status  |
| ------------- | ------------- | ------------- | ------------- |
|  Y  |  Y  |  Y  |  OK  |
|  Y  |  Y  |  N  |  Error  |
|  Y  |  N  |  Y  |  Error  |
|  Y  |  N  |  N  |  OK  |
|  N  |  Y  |  Y  |  OK  |
|  N  |  Y  |  N  |  Error  |
|  N  |  N  |  Y  |  Error  |
|  N  |  N  |  N  |  Error  |

Reply:
Thanks for the summary. I believe that the corrected text reflects the rules of that table.